### PR TITLE
KMP Algorithm

### PR DESCRIPTION
**Pull request by vinaya563**
## Related Issue
- Issue number (#1238 )

## Proposed Changes
- Added the implementation and required readme for the KMP Algorithm which is one of the most important string algorithms used for pattern searching.

## Checklist

- [x] ✅ My code follows the code style of this project.
- [x] 📝 My change requires a change to the README.
- [x] 🎀 I have updated the README accordingly.
- [x] 👀 I have read the [**CONTRIBUTING**](https://github.com/Lakhankumawat/LearnCPP/blob/main/Contributing.md) document.
- [x] ✨ I have added tests to cover my changes.
- [x] 🚩 All new and existing tests passed.
- [x] 🌟 ed the repo

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Output Screenshots
image

### CODE REVIEW

**Comment by Mrjoy832** [comment]
Use ### for Sub-headings of readme

**Comment by Mrjoy832** [comment]
Take String and key as input from user

**Comment by vinaya563** [comment]
@Mrjoy832  I have done the necessary changes ..Can you please review it now

**Comment by Mrjoy832** [comment]
LGTM🌟

File: S-StringAlgorithms/KMP/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+
+<!-- Table of content -->
+### Table of content
+-[Problem_statement](#problemstatement)
+-[KMP_Algorithm](#kmpalgorithm)
+-[Algorithm](#algorithm)
+-[Complexity_Analysis](#complexityanalysis)
+
```
Comment (Lakhankumawat) [inline S-StringAlgorithms/KMP/README.md:8]: Correct Links 
1. Some of them are not pointing the right thing
2. They are not line by line unlike others 
image


## Tip :
do this `- []` space between `-` and `[brackets]`

File: S-StringAlgorithms/KMP/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+### Algorithm
+
+Unlike Naive algorithm, where we slide the pattern by one and compare all characters at each shift, we use a value from lps[] to decide the next characters to be matched. The idea is to not match a character that we know will anyway match.
+
+How to use lps[] to decide next positions (or to know a number of characters to be skipped)?
+
+We start comparison of pat[j] with j = 0 with characters of current window of text.
+We keep matching characters txt[i] and pat[j] and keep incrementing i and j while pat[j] and txt[i] keep matching.
+When we see a mismatch
+We know that characters pat[0..j-1] match with txt[i-j…i-1] (Note that j starts with 0 and increment it only when there is a match).
```
Comment (Lakhankumawat) [inline S-StringAlgorithms/KMP/README.md:35]: If this is pseudocode then add in it ` "```" ` section
```
Like this is a pseudo code
```

File: S-StringAlgorithms/KMP/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+We also know (from above definition) that lps[j-1] is count of characters of pat[0…j-1] that are both proper prefix and suffix.
+From above two points, we can conclude that we do not need to match these lps[j-1] characters with txt[i-j…i-1] because we know that these characters will anyway match. Let us consider above example to understand this.
+ ![kmped](https://user-images.githubusercontent.com/91210199/164911136-c063849b-65e3-4c60-9235-cdd64367b0f7.jpg)
+
+
+
+### Complexity Analysis
+
+Time Complexity: O(n) in the worst case where n is the length of text.
+Space Complexity: O(m) where m is the size of the pattern.
```
Comment (Lakhankumawat) [inline S-StringAlgorithms/KMP/README.md:45]: Advantages and Disadv. are missing

You need to follow particular guidelines for your README , please follow
these guidelines [click here](https://github.com/Lakhankumawat/LearnCPP#to-add-programs-using-stl-follow-same-approach-as--algorithms) 
also you can see how your README should look like [see here](https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md) 
for more please go through these [here](https://youtu.be/7UgC3HDETRA)

For god sake please follow guidelines i've told you same earlier in your PR's as well.

**Review by Lakhankumawat — CHANGES_REQUESTED** [review]

**Comment by vinaya563** [comment]
@Lakhankumawat
 sorry for inconvenience sir...I assure you that I will not repeat the same mistakes again.
Now is it okay ??

**Review by Lakhankumawat — APPROVED** [review]
### LGTM ✨🎉

**Comment by Lakhankumawat** [comment]
Good Contribution .

**Pull request merged**